File: 卡码网/Python/test22.py
#22. 二叉树的遍历
class Node:
	nodes=[-1]
	def __init__(self,data,left,right):
		self.data=data
		self.left=left
		self.right=right
	def preOrder(self):
		print(self.data,end="")
		if self.left!=0:
			Node.nodes[self.left].preOrder()
		if self.right!=0:
			Node.nodes[self.right].preOrder()
	def inOrder(self):
		if self.left!=0:
			Node.nodes[self.left].inOrder()
		print(self.data,end="")
		if self.right!=0:
			Node.nodes[self.right].inOrder()
	def postOrder(self):
		if self.left!=0:
			Node.nodes[self.left].postOrder()
		if self.right!=0:
			Node.nodes[self.right].postOrder()
		print(self.data,end="")

def main():
	n=int(input())
	for i in range(1,n+1):
		line=input().split()
		left,right=int(line[1]),int(line[2])
		Node.nodes.append(Node(line[0],left,right))
	Node.nodes[1].preOrder()
	print()
	Node.nodes[1].inOrder()
	print()
	Node.nodes[1].postOrder()

if __name__ == '__main__':
	main()

File: 卡码网/C++/test22.cpp
//22. 二叉树的遍历
#include <iostream>
#include <vector>
using namespace std;

struct Node{
	char data;
	int left,right;
	// Node(char data,int left,int right):data(data),left(left),right(right){}
};

vector<Node> nodes(1);

void preOrder(Node root){
	printf("%c",root.data);
	if(root.left!=0) preOrder(nodes[root.left]);
	if(root.right!=0) preOrder(nodes[root.right]);
}

void inOrder(Node root){
	if(root.left!=0) inOrder(nodes[root.left]);
	printf("%c",root.data);
	if(root.right!=0) inOrder(nodes[root.right]);
}

void postOrder(Node root){
	if(root.left!=0) postOrder(nodes[root.left]);
	if(root.right!=0) postOrder(nodes[root.right]);
	printf("%c",root.data);
}

int main(){
	int n;
	scanf("%d",&n);
	char data;
	int left,right;
	for(int i=1;i<=n;i++){
		getchar();
		scanf("%c %d %d",&data,&left,&right);
		nodes.push_back(Node{data,left,right});
	}
	preOrder(nodes[1]);
	printf("\n");
	inOrder(nodes[1]);
	printf("\n");
	postOrder(nodes[1]);
	return 0;
}